source/common/upstream/thread_aware_lb_impl.cc

```diff
@@ -185,5 +186,108 @@ LoadBalancerPtr ThreadAwareLoadBalancerBase::LoadBalancerFactoryImpl::create() {
+  /*
+  Consistent Hashing with Bounded Load
+  Ref: https://arxiv.org/abs/1608.01350
+  */
+  uint32_t overall_active = host->cluster().stats().upstream_rq_active_.value();
+  uint32_t host_active = host->stats().rq_active_.value();
+
+  uint32_t total_slots = ((overall_active + 1) * hash_balance_factor_ + 99) / 100;
+  uint32_t slots =
+      std::max(static_cast<uint32_t>(std::ceil(total_slots * weight)), static_cast<uint32_t>(1));
+
+  if (host->stats().rq_active_.value() > slots) {
+    ENVOY_LOG_MISC(
+        debug,
+        "ThreadAwareLoadBalancerBase::BoundedLoadHashingLoadBalancer::chooseHost: "
+        "host {} overloaded; overall_active {}, host_weight {}, host_active {} > slots {}",
+        host->address()->asString(), overall_active, weight, host_active, slots);
+    return true;
+  }
+  return false;
+}
+
+HostConstSharedPtr
+ThreadAwareLoadBalancerBase::BoundedLoadHashingLoadBalancer::chooseHost(uint64_t hash,
+                                                                        uint32_t attempt) const {
+
+  if (hlb_ptr_ == nullptr) {
+    return nullptr;
+  }
+  if (normalized_host_weights_ == nullptr || normalized_host_weights_->empty()) {
+    return nullptr;
+  }
+
+  HostConstSharedPtr host = hlb_ptr_->chooseHost(hash, attempt);
+  if (host == nullptr) {
+    return nullptr;
+  }
+  const double weight = normalized_host_weights_map_.at(host);
+  if (!isHostOverloaded(host, weight)) {
+    ENVOY_LOG_MISC(debug,
+                   "ThreadAwareLoadBalancerBase::BoundedLoadHashingLoadBalancer::chooseHost: "
+                   "selected host #{} (attempt:1)",
+                   host->address()->asString());
+    return host;
+  }
+
+  // Revisiting Consistent Hashing with Bounded Loads
+  // https://arxiv.org/abs/1908.08762
+
+  // When a host is overloaded, we choose the next host in a random manner rather than picking the
+  // next one in the ring. The random sequence is seeded by the hash, so the same input gets the
+  // same sequence of hosts all the time.
+  const uint32_t num_hosts = normalized_host_weights_->size();
+  auto host_index = std::vector<uint32_t>(num_hosts);
+  for (uint32_t i = 0; i < num_hosts; i++) {
+    host_index[i] = i;
+  }
+
+  // Not using Random::RandomGenerator as it does not take a seed. Seeded RNG is a requirement
+  // here as we need the same shuffle sequence for the same hash every time.
+  // Further, not using std::default_random_engine and std::uniform_int_distribution as they
+  // are not consistent across Linux and Windows platforms.
+  const uint64_t seed = hash;
+  std::mt19937 random(seed);
+
+  // generates a random number in the range [0,k) uniformly.
+  auto uniform_int = [](std::mt19937 random, uint32_t k) -> uint32_t {
+    // Note: 1+random()%k is biased
+    uint32_t x = k;
+    while (x >= k) {
+      x = random() / ((static_cast<uint64_t>(random.max()) + 1u) / k);
+    }
+    return x;
+  };
+
+  HostConstSharedPtr h;
+  for (uint32_t i = 0; i < num_hosts; i++) {
```

Comment:
Is there any situation here in which we would want to limit the iterations and potentially return the best thing we found so far?

Reply:
This seems to take us from O(1) to O(n), which might surprise some users.

Reply:
@mattklein123 limiting the iterations would affect the guarantee on the bounded load. We can potentially have another variant of consistent hash load balancing where every hash is mapped to k hosts instead of 1. However, I dont know how this will work with weighted hosts. If performance is a concern the users should use a higher ```hash_balance_factor```, however, this will have higher load on some hosts. I mentioned this in the documentation of ```hash_balance_factor``` config parameter.